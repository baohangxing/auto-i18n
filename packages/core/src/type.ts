export interface I18nCallRule {
  /** caller of functionName
       *  @default ''
      */
  caller: string
  /** function name of `t`
       *  @default 't'
      */
  functionName: string

  definedDeclaration: string
  /**
       * import i18n t Declaration
       */
  importDeclaration: string
}

export type FileExtension = 'js' | 'ts' | 'cjs' | 'mjs' | 'jsx' | 'tsx' | 'vue'

export interface AutoConfig {
  /**
      * local Locales-Json directory Paths, all user json file should found in this paths
      *
      * @example [`/Users/XXX/GM/src/lang/locales`]
      */
  localesJsonDirs: string[] | string

  /**
       * locales used, the locale will not trans if not in `locales`
       *
       * @example [`ja-jp`, `zh-cn`]
       */
  locales: string[]

  /**
       * the base locale in your project, make sure it is in `locales` and all words has configured in locale-Json file
       *
       * @default `zh-cn`
       */
  baseLocale: string

  /**
       *
       * @param locale
       * @returns
       */
  untransSymbol: (locale: string) => string

  /**
 * translate word rule when not translated
 *
 * @param word
 * @param locale the word current locale, your baseLocale
 * @param toLocale to translate locale like `ja-jp`
 * @returns
 */
  transLacaleWord?: (word: string, locale: string, toLocale: string) => Promise<string>

  /**
      * files to auto trans, using fast-glob rule
      *
      * see deteils: https://github.com/mrmlnc/fast-glob
      *
      * @example ["src/*.{vue}", "view/**"]
      */
  includes: string[]

  /**
       * exclude files to auto trans, using fast-glob rule
       *
       * see deteils: https://github.com/mrmlnc/fast-glob
       *
       * @example  ['**\/node_modules/**\/*']
       */
  exclude: string[]

  /**
       * output filed directory path,
       * the default output filed directory path is your project path
       *
       * @default `./ `
       */
  outputFileDir: string

  i18nCallRules: Record<FileExtension, I18nCallRule>
}

export interface TransCommandOption {
  filePath: string
  generateNewFile: boolean
  newFileName: string
}

export interface LangJson {
  name: string
  path: string
}
