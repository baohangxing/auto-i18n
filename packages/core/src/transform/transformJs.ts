import { createRequire } from 'module'
import type { NodePath } from '@babel/traverse'
import type {
  ArrayExpression,
  CallExpression,
  Comment,
  Expression,
  ImportDeclaration,
  JSXAttribute,
  JSXText,
  MemberExpression,
  ObjectExpression,
  ObjectProperty,
  Program,
  StringLiteral,
  TemplateLiteral,
} from '@babel/types'
import type { GeneratorResult } from '@babel/generator'
import template from '@babel/template'
import { isEmpty, isObject } from 'lodash-es'
import t from '@babel/types'
import type * as traverseType from '@babel/traverse/index'
import type * as generatorType from '@babel/generator/index'
import type { ParseResult } from '@babel/core'
import type { Collector, I18nCallRule, Log } from '../types'
import { includeChinese } from '../utils/help'
import { IGNORE_REMARK } from '../config/constants'
import { getAutoBaseConfig as getAutoConfig } from '../config/config'
import { escapeQuotes, getCallExpression, getStringLiteral } from './tools'

const require = createRequire(import.meta.url)
// see https://github.com/babel/babel/issues/13855
const traverse: typeof traverseType.default = require('@babel/traverse').default
const babelGenerator: typeof generatorType.default = require('@babel/generator').default

interface NamedInterpolationParams {
  [k: string]:
  | string
  | {
    isAstNode: true
    value: Expression
  }
}

type ListInterpolationParams = Array< string
| {
  isAstNode: true
  value: Expression
}>

interface TransformJsOptions {
  rule: I18nCallRule
  parse: (code: string) => ParseResult | null

  replace: boolean

  /**
   * Whether handle js in vue
   */
  isJsInVue?: boolean

  loger?: Log<any>

  collector: Collector
}

const getObjectExpression = (obj: NamedInterpolationParams): ObjectExpression => {
  const ObjectPropertyArr: ObjectProperty[] = []
  Object.keys(obj).forEach((k) => {
    const tempValue = obj[k]
    let newValue: Expression
    if (isObject(tempValue))
      newValue = tempValue.value
    else
      newValue = t.identifier(tempValue)

    ObjectPropertyArr.push(t.objectProperty(t.identifier(k), newValue))
  })
  const ast = t.objectExpression(ObjectPropertyArr)
  return ast
}

const getArrayExpression = (params: ListInterpolationParams): ArrayExpression => {
  const ObjectPropertyArr: Expression [] = []
  params.forEach((tempValue) => {
    let newValue: Expression
    if (isObject(tempValue))
      newValue = tempValue.value
    else
      newValue = t.identifier(tempValue)

    ObjectPropertyArr.push(newValue)
  })
  const ast = t.arrayExpression(ObjectPropertyArr)
  return ast
}

/**
 * 判断节点是否是vue组件props属性的默认值，例如：
 * ```
 * props: {
    title: {
      default: '标题'
    }
  }
 * ```
 */
const isPropDefaultStringLiteralNode = (path: NodePath<StringLiteral>): boolean => {
  const objWithProps = path.parentPath?.parentPath?.parentPath?.parentPath?.parent
  const rootNode
    = path.parentPath?.parentPath?.parentPath?.parentPath?.parentPath?.parentPath?.parent
  let isMeetProp = false
  let isMeetKey = false
  let isMeetContainer = false
  // 属性是否包含在props结构里
  if (
    objWithProps
    && objWithProps.type === 'ObjectProperty'
    && objWithProps.key.type === 'Identifier'
    && objWithProps.key.name === 'props'
  )
    isMeetProp = true

  // 对应key是否是default
  if (
    path.parent
    && path.parent.type === 'ObjectProperty'
    && path.parent.key.type === 'Identifier'
    && path.parent.key.name === 'default'
  )
    isMeetKey = true

  // 遍历到指定层数后是否是导出声明
  if (rootNode && rootNode.type === 'ExportDefaultDeclaration')
    isMeetContainer = true

  return isMeetProp && isMeetKey && isMeetContainer
}

const getReplaceValue = (rule: I18nCallRule, value: string, replaceValue: string,
  params?: NamedInterpolationParams | ListInterpolationParams,
): Expression => {
  value = escapeQuotes(value)
  const { transCaller, transIdentifier } = rule
  let expression: string
  if (params) {
    const keyLiteral = getStringLiteral(replaceValue)
    if (transCaller) {
      return t.callExpression(
        t.memberExpression(t.identifier(transCaller), t.identifier(transIdentifier)),
        [keyLiteral, !Array.isArray(params)
          ? getObjectExpression(params)
          : getArrayExpression(params),
        ],
      )
    }
    else {
      return t.callExpression(t.identifier(transIdentifier), [
        keyLiteral,
        !Array.isArray(params)
          ? getObjectExpression(params)
          : getArrayExpression(params),
      ])
    }
  }
  else {
    expression = getCallExpression(rule, replaceValue)
    return template.expression(expression)()
  }
}

const transformJs = (code: string, options: TransformJsOptions): GeneratorResult => {
  const autoConfig = getAutoConfig()

  const rule = options.rule
  // 文件是否导入过i18n
  let hasImportI18n = false

  // 文件里是否存在中文转换，有的话才有必要导入i18n
  let hasTransformed = false

  const transformAST = (code: string, options: TransformJsOptions) => {
    const collector = options.collector

    function getTraverseOptions() {
      return {
        enter(path: NodePath) {
          const leadingComments = path.node.leadingComments
          if (leadingComments) {
            // 是否跳过翻译
            let isSkipTransform = false
            leadingComments.every((comment: Comment) => {
              if (comment.value.includes(IGNORE_REMARK)) {
                isSkipTransform = true
                return false
              }
              return true
            })
            if (isSkipTransform)
              path.skip()
          }
        },

        StringLiteral(path: NodePath<StringLiteral>) {
          if (includeChinese(path.node.value) && options.isJsInVue && isPropDefaultStringLiteralNode(path)) {
            collector.add(path.node.value)
            if (options.replace) {
              const expression = `() => ${getCallExpression(rule, collector.getKey(path.node.value))}`
              path.replaceWith(template.expression(expression)())
              hasTransformed = true
            }
            path.skip()
            return
          }

          if (includeChinese(path.node.value)) {
            collector.add(path.node.value)
            if (options.replace) {
              path.replaceWith(
                getReplaceValue(rule, path.node.value, collector.getKey(path.node.value)),
              )
              hasTransformed = true
            }
          }
          path.skip()
        },

        TemplateLiteral(path: NodePath<TemplateLiteral>) {
          const { node } = path
          const templateMembers = [...node.quasis, ...node.expressions]
          templateMembers.sort((a, b) => (a.start as number) - (b.start as number))

          const shouldReplace = node.quasis.some(node => includeChinese(node.value.raw))

          if (shouldReplace) {
            let value = ''
            let slotParams: NamedInterpolationParams | ListInterpolationParams | undefined

            if (autoConfig.transInterpolationsMode === 'NamedInterpolationMode') {
              let slotIndex = 1
              const params: NamedInterpolationParams = {}
              templateMembers.forEach((node) => {
                if (node.type === 'Identifier') {
                  value += `{${node.name}}`
                  params[node.name] = node.name
                }
                else if (node.type === 'TemplateElement') {
                  value += node.value.raw.replace(/[\r\n]/g, '') // 用raw防止字符串中出现 /n
                }
                else if (node.type === 'MemberExpression') {
                  const key = `slot${slotIndex++}`
                  value += `{${key}}`
                  params[key] = {
                    isAstNode: true,
                    value: node as MemberExpression,
                  }
                }
                else {
                // 处理${}内容为表达式的情况。例如`测试${a + b}`，
                // 把 a + b 这个语法树作为params的值, 并自定义params的键为slot加数字的形式
                  const key = `slot${slotIndex++}`
                  value += `{${key}}`
                  const expression = babelGenerator(node).code
                  const tempAst = transformAST(expression, options) as any
                  const expressionAst = tempAst.program.body[0].expression
                  params[key] = {
                    isAstNode: true,
                    value: expressionAst,
                  }
                }
              })

              slotParams = isEmpty(params) ? undefined : params
            }
            else if (autoConfig.transInterpolationsMode === 'ListInterpolationMode') {
              let index = 0
              const params: ListInterpolationParams = []
              templateMembers.forEach((node) => {
                if (node.type === 'Identifier') {
                  value += `{${index++}}`
                  params.push(node.name)
                }
                else if (node.type === 'TemplateElement') {
                  value += node.value.raw.replace(/[\r\n]/g, '')
                  // TODO 用raw防止字符串中出现 /n??
                }
                else if (node.type === 'MemberExpression') {
                  value += `{${index++}}`
                  params.push({
                    isAstNode: true,
                    value: node as MemberExpression,
                  })
                }
                else {
                  value += `{${index++}}`
                  const expression = babelGenerator(node).code
                  const tempAst = transformAST(expression, options) as any
                  const expressionAst = tempAst.program.body[0].expression
                  params.push({
                    isAstNode: true,
                    value: expressionAst,
                  })
                }
              })

              slotParams = isEmpty(params) ? undefined : params
            }
            else {
              options.loger?.error('AutoConfig.transInterpolationsMode must is '
                + '\'NamedInterpolationMode\' or \'ListInterpolationMode\'')
            }

            collector.add(value)

            if (options.replace) {
              path.replaceWith(getReplaceValue(rule, value, collector.getKey(value), slotParams))
              hasTransformed = true
            }
          }
        },

        JSXText(path: NodePath<JSXText>) {
          if (includeChinese(path.node.value)) {
            collector.add(path.node.value)
            if (options.replace) {
              // TODO trim
              path.replaceWith(t.jSXExpressionContainer(
                getReplaceValue(rule, path.node.value, collector.getKey(path.node.value))))
              hasTransformed = true
            }
          }
          path.skip()
        },

        JSXAttribute(path: NodePath<JSXAttribute>) {
          const node = path.node as NodePath<JSXAttribute>['node']
          const valueType = node.value?.type
          if (valueType === 'StringLiteral' && node.value && includeChinese(node.value.value)) {
            collector.add(node.value.value)
            if (options.replace) {
              const jsxIdentifier = t.jsxIdentifier(node.name.name as string)
              const jsxContainer = t.jSXExpressionContainer(
                getReplaceValue(rule, node.value.value, collector.getKey(node.value.value)),
              )
              path.replaceWith(t.jsxAttribute(jsxIdentifier, jsxContainer))
              hasTransformed = true
            }
            path.skip()
          }
        },

        CallExpression(path: NodePath<CallExpression>) {
          const { node } = path
          const { transCaller, transIdentifier } = rule
          const callee = node.callee

          // 无调用对象的情况，例如 t('xx')
          if (!transCaller && callee.type === 'Identifier' && callee.name === transIdentifier) {
            path.skip()
            return
          }

          // 有调用对象的情况，例如this.$t('xx')、i18n.$t('xx')
          if (callee.type === 'MemberExpression') {
            if (callee.property && callee.property.type === 'Identifier') {
              if (callee.property.name === transIdentifier) {
                // 处理形如i18n.$t('xx')的情况
                if (callee.object.type === 'Identifier' && callee.object.name === transCaller) {
                  path.skip()
                  return
                }
                // 处理形如this.$t('xx')的情况
                if (callee.object.type === 'ThisExpression' && transCaller === 'this')
                  path.skip()
              }
            }
          }
        },

        ImportDeclaration(path: NodePath<ImportDeclaration>) {
          const { importDeclaration, variableDeclaration } = rule
          const res = importDeclaration.match(/from ["'](.*)["']/)
          const packageName = res ? res[1] : ''

          if (path.node.source.value === packageName)
            hasImportI18n = true

          if (!hasImportI18n && hasTransformed) {
            const program = path.parent as Program
            const lastImportIndex = program.body.findIndex(x => x.type !== 'ImportDeclaration')

            const variableAst = template.statements(variableDeclaration)().reverse()

            variableAst.forEach((statement) => {
              program.body.splice(lastImportIndex, 0, statement)
            })

            const importAst = template.statements(importDeclaration)().reverse()

            importAst.forEach((statement) => {
              program.body.splice(lastImportIndex, 0, statement)
            })

            hasImportI18n = true
          }
        },
      }
    }

    const ast = options.parse(code)
    traverse(ast, getTraverseOptions())
    return ast
  }

  const ast = transformAST(code, options)

  if (!hasImportI18n && hasTransformed && ast) {
    const { importDeclaration, variableDeclaration } = rule
    const program = ast.program
    const lastImportIndex = program.body.findIndex(x => x.type !== 'ImportDeclaration')

    const variableAst = template.statements(variableDeclaration)().reverse()

    variableAst.forEach((statement) => {
      program.body.splice(lastImportIndex, 0, statement)
    })

    const importAst = template.statements(importDeclaration)().reverse()

    importAst.forEach((statement) => {
      program.body.splice(lastImportIndex, 0, statement)
    })
    hasImportI18n = true
  }

  const result = (!ast ? code : babelGenerator(ast)) as GeneratorResult

  return result
}

export default transformJs

export type { TransformJsOptions }
