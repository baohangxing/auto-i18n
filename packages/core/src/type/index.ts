export * from './command'
export * from './autoConfig'
