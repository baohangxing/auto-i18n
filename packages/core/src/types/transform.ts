import type { ParseResult } from '@babel/core'

export interface I18nCallRule {
  /** transCaller of transIdentifier
     *  @default ''
     */
  transCaller: string

  /** function name of `t`
     *  @default 't'
     */
  transIdentifier: string

  variableDeclaration: string

  /**
    * import i18n t Declaration
    */
  importDeclaration: string
}

export interface transformOptions {
  rule: I18nCallRule
  parse: (code: string) => ParseResult | null

  /** handle js in vue */
  isJsInVue?: boolean
}
