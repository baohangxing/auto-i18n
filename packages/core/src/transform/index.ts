import type { FileExtension, I18nCallRules } from '../types'
import log from '../utils/log'
import transformJs from './transformJs'
import transformVue from './transformVue'
import { initJsParse, initTsxParse } from './parse'

const transform = (
  code: string,
  ext: FileExtension,
  rules: I18nCallRules,
  replace = true,
): {
  code: string
} => {
  switch (ext) {
    case 'cjs':
    case 'mjs':
    case 'js':
    case 'jsx':
      return transformJs(code, {
        rule: rules[ext],
        parse: initJsParse(),
      }, replace)
    case 'ts':
    case 'tsx':
      return transformJs(code, {
        rule: rules[ext],
        parse: initTsxParse(),
      }, replace)
    case 'vue':
      return transformVue(code, rules[ext], replace)
    default:
      log.error(`not support transform .${ext} extension`)
      return { code }
  }
}

export { transform }
