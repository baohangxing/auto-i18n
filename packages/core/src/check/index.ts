import { createRequire } from 'module'
import type { NodePath } from '@babel/traverse'
import type {
  CallExpression,
} from '@babel/types'
import type * as traverseType from '@babel/traverse/index'
import { parse } from '@vue/compiler-sfc'
import * as htmlparser2 from 'htmlparser2'
import mustache from 'mustache'
import type { ParseResult } from '@babel/core'
import type { I18nCallRule, Log } from '../types'
import { IGNORE_REMARK } from '../config/constants'
import { trimValue } from '../transform/tools'
import { initTsxParse } from '../transform/parse'

interface GetI18nKeysOptions {
  rule: I18nCallRule
  parse: (code: string) => ParseResult | null

  logger?: Log<any>
}

const require = createRequire(import.meta.url)
// see https://github.com/babel/babel/issues/13855
const traverse: typeof traverseType.default = require('@babel/traverse').default

const getJsI18nKeys = (code: string, options: GetI18nKeysOptions): string[] => {
  const rule = options.rule

  const res = new Set<string>()

  const checkAST = (code: string, options: GetI18nKeysOptions) => {
    function getTraverseOptions() {
      return {

        CallExpression(path: NodePath<CallExpression>) {
          const { node } = path
          const { transCaller, transIdentifier } = rule
          const callee = node.callee

          let mayIsArguments = false

          // 无调用对象的情况，例如 t('xx')
          if (!transCaller && callee.type === 'Identifier' && callee.name === transIdentifier)
            mayIsArguments = true

          // 有调用对象的情况，例如this.$t('xx')、i18n.$t('xx')
          if (callee.type === 'MemberExpression') {
            if (callee.property && callee.property.type === 'Identifier') {
              if (callee.property.name === transIdentifier) {
                // 处理形如i18n.$t('xx')的情况
                if (callee.object.type === 'Identifier' && callee.object.name === transCaller)
                  mayIsArguments = true

                // 处理形如this.$t('xx')的情况
                if (callee.object.type === 'ThisExpression' && transCaller === 'this')
                  mayIsArguments = true
              }
            }
          }

          if (mayIsArguments && node.arguments[0]) {
            if (node.arguments[0].type === 'StringLiteral')
              res.add(node.arguments[0].value)

            if (node.arguments[0].type === 'TemplateLiteral'
              && node.arguments[0].expressions.length === 0
              && node.arguments[0].quasis?.[0].value.raw !== '')
              res.add(node.arguments[0].quasis[0].value.raw)
          }
        },
      }
    }

    const ast = options.parse(code)
    traverse(ast, getTraverseOptions())
    return ast
  }

  checkAST(code, options)

  return [...res]
}

const getJsSyntaxI18nKeys = (source: string, options: GetI18nKeysOptions): string[] => {
  if (source.startsWith('{') && source.endsWith('}'))
    source = `___temp = ${source}`

  return getJsI18nKeys(source,
    {
      rule: {
        ...options.rule,
        variableDeclaration: '',
        importDeclaration: '',
      },
      parse: initTsxParse(),
      logger: options.logger,
    },
  )
}

const getTemplateKeys = (code: string, options: GetI18nKeysOptions): string[] => {
  const res: string [] = []

  let shouldIgnore = false
  const parser = new htmlparser2.Parser(
    {
      onopentag(tagName, attributes) {
        if (shouldIgnore)
          return

        for (const key in attributes) {
          const attrValue = attributes[key]
          // v-for 的 vue自定义语法无法用 getJsSyntaxI18nKeys
          const isVueDirective = key.startsWith(':') || key.startsWith('@')
            || (key.startsWith('v-') && !key.startsWith('v-for'))
          if (isVueDirective) {
            const keys = getJsSyntaxI18nKeys(attrValue, options)
            res.push(...keys)
          }
        }
      },

      ontext(text) {
        if (shouldIgnore)
          return

        const tokens = mustache.parse(text)
        for (const token of tokens) {
          const type = token[0]
          let value = token[1]

          value = trimValue(value)
          if (type === 'name') {
            const keys = getJsSyntaxI18nKeys(value, options)
            res.push(...keys)
          }
        }
      },
      onclosetag() {
        shouldIgnore = false
      },
      oncomment(comment) {
        if (comment.includes(IGNORE_REMARK))
          shouldIgnore = true
      },
    },
    {
      lowerCaseTags: false,
      recognizeSelfClosing: true,
      lowerCaseAttributeNames: false,
    },
  )

  parser.write(code)
  parser.end()
  return res
}

const getScriptKeys = (source: string, options: GetI18nKeysOptions): string[] => {
  return getJsI18nKeys(source, options)
}

const getVueI18nKeys = (
  code: string,
  options: GetI18nKeysOptions,
): string[] => {
  const { descriptor, errors } = parse(code)
  if (errors.length > 0) {
    options.logger?.error('Parse vue error', errors)
    return []
  }

  const res = new Set<string>()

  const { template, script, scriptSetup } = descriptor

  if (template) {
    const keys = getTemplateKeys(template.content, options)
    for (const x of keys)
      res.add(x)
  }

  if (script) {
    const keys = getScriptKeys(script.content, options)
    for (const x of keys)
      res.add(x)
  }

  if (scriptSetup) {
    const keys = getScriptKeys(scriptSetup.content, options)
    for (const x of keys)
      res.add(x)
  }

  return [...res]
}

export { getJsI18nKeys, getVueI18nKeys }
export type { GetI18nKeysOptions }
