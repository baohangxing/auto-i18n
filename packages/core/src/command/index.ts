import { Command, Option } from 'commander'
import { CLI_CONFIG_NAME } from '../config/constants'
import type { RevertCommandOption, TransCommandOption } from '../types/config'
import { checkAllTranslated } from '../check'
import { revert } from '../revert'
import { version } from './../../package.json'
import { generateXlsx } from './genXlsx'
import { init } from './init'
import { trans } from './trans'
import { updateLocales, updateLocalesFromXlsx } from './update'

const program = new Command()

program
  .name(CLI_CONFIG_NAME)
  .description('A CLI to help you transform your project '
    + 'to an internationalization project automatically.')
  .version(version)

program.addOption(new Option('-d, --debug').hideHelp())
program.addOption(new Option('-vb, --verbose').hideHelp())

program.on('option:verbose', () => {
  process.env.AUTO_I18N_VERBOSE = program.opts().verbose
})
program.on('option:debug', () => {
  process.env.AUTO_I18N_DEBUG = program.opts().debug
})

program.command('init')
  .description('Init yo-auto-i18n, generate an new default config file automatically.')
  .action(() => {
    init()
  })

program.command('trans')
  .description('Transform a single file or all files.')
  .argument('[transPath]', 'file or directory path')
  .option('--modify', 'whether to modify on the source file')
  .option('-t, --template <string>', 'template json file which inclues all language key')
  .action((transPath, options) => {
    const transCommandOption: TransCommandOption = {
      transPath: transPath ?? '',
      modifyMode: !!options.modify,
      templateFile: options.template ?? '',
    }
    trans(transCommandOption)
  })

program.command('update')
  .description('Update other locales Json files by base locales Json file')
  .option('-t, --templateXlsx <string>', 'update locales json files by the Xlsx template file,'
    + ' the frirt line of the first sheet should have all locales names and the `key`', '')
  .action(async (options) => {
    await updateLocales()
    options.templateXlsx && (await updateLocalesFromXlsx(options.templateXlsx))
  })

program.command('genlsx')
  .description('Generate an Xlsx file by locales json files,'
    + ' the frirt line of in the first sheet of the xlsx will have all locales-name and the `key`')
  .action(() => {
    generateXlsx()
  })

program.command('check')
  .description('Check weather all word has been transformed or not')
  .action(() => {
    checkAllTranslated()
  })

program.command('revert')
  .description('revert a transformed file or files in a directory')
  .argument('<revertPath>', 'file or directory path')
  .option('-t, --target <string>', 'the target language when revert')
  .action(async (revertPath, options) => {
    const revertCommandOption: RevertCommandOption = {
      revertPath: revertPath ?? '',
      target: options.target ?? '',
    }
    await revert(revertCommandOption)
  })

program.parse()
