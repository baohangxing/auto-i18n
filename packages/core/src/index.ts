import './command'
export * from './types'
