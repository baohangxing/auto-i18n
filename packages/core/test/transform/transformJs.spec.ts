import { describe, expect, it } from 'vitest'
import type { AutoBaseConfig, Collector } from '../../src/types'
import { initJsParse } from '../../src/transform'
import type { TransformJsOptions } from '../../src/transform/transformJs'
import transformJs from '../../src/transform/transformJs'
import { defaultAutoBaseConfig } from '../../dist'

const genTransOptions = (collector: Collector, autoConfig?: Partial<AutoBaseConfig>): TransformJsOptions => {
  return {
    rule: {
      transCaller: '',
      transIdentifier: 't',
      variableDeclaration: 'const t = useI18n()',
      importDeclaration: 'import { useI18n } from \'vue-i18n\'',
    },
    parse: initJsParse(),
    autoConfig: Object.assign({}, defaultAutoBaseConfig, autoConfig),
    replace: true,
    collector,
  }
}

describe('#transform', () => {
  it('should transform basic code', async () => {
    const words: string[] = []
    const reslut = transformJs(`
    let title = "标题"
    `, genTransOptions(
      {
        add(str) {
          words.push(str)
        },
        getKey(str: string) {
          return `Key of ${str}`
        },
      }),
    )
    expect(words).toEqual(['标题'])
    expect(reslut.code).toMatchSnapshot()
  })

  it('should transform code when import', async () => {
    const words: string[] = []
    const reslut = transformJs(`
    import { useI18n } from \'vue-i18n\'
    const t = useI18n()
    
    let x = t('s')
    let title = \`标题\`
    `, genTransOptions(
      {
        add(str) {
          words.push(str)
        },
        getKey(str: string) {
          return `Key of ${str}`
        },
      }),
    )
    expect(words).toEqual(['标题'])
    expect(reslut.code).toMatchSnapshot()
  })

  it('should transform code in function', async () => {
    const words: string[] = []
    const reslut = transformJs(`
    import fs from 'fs'
    let title = console.log('标题')
    `, genTransOptions(
      {
        add(str) {
          words.push(str)
        },
        getKey(str: string) {
          return `Key of ${str}`
        },
      }),
    )
    expect(words).toEqual(['标题'])
    expect(reslut.code).toMatchSnapshot()
  })

  it('should transform code in template', async () => {
    const words: string[] = []
    const reslut = transformJs(
      // eslint-disable-next-line no-template-curly-in-string
      'const title = `标题\$\{msg\}`'
      , genTransOptions(
        {
          add(str) {
            words.push(str)
          },
          getKey(str: string) {
            return `Key of ${str}`
          },
        }),
    )
    expect(words).toEqual(['标题{msg}'])
    expect(reslut.code).toMatchSnapshot()
  })

  it('should transform code', async () => {
    const words: string[] = []
    const reslut = transformJs(
      // eslint-disable-next-line no-template-curly-in-string
      'const title = `标题\$\{msg\}`'
      , genTransOptions(
        {
          add(str) {
            words.push(str)
          },
          getKey(str: string) {
            return `Key of ${str}`
          },
        }, {
          transInterpolationsMode: 'ListInterpolationMode',
        }),
    )
    expect(words).toEqual(['标题{0}'])
    expect(reslut.code).toMatchSnapshot()
  })

  it('should transform jsx code', async () => {
    const words: string[] = []
    const reslut = transformJs(
      'const title = (<div>不好不好{aa}</div>)'
      , genTransOptions(
        {
          add(str) {
            words.push(str)
          },
          getKey(str: string) {
            return `Key of ${str}`
          },
        }, {
          transInterpolationsMode: 'ListInterpolationMode',
        }),
    )
    expect(words).toEqual(['不好不好'])
    expect(reslut.code).toMatchSnapshot()
  })

  it('should don\'t transform code if ignored', async () => {
    const words: string[] = []
    transformJs(
      `//auto-i18n-ignore
      let title = "标题" 
      `
      , genTransOptions(
        {
          add(str) {
            words.push(str)
          },
          getKey(str: string) {
            return `Key of ${str}`
          },
        }, {
          transInterpolationsMode: 'ListInterpolationMode',
        }),
    )
    expect(words).toEqual([])
  })
})
