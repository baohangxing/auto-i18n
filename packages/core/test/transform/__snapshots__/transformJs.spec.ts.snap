// Vitest Snapshot v1

exports[`#transform > should transform basic code 1`] = `
"import { useI18n } from 'vue-i18n';
const t = useI18n();
let title = t('Key of 标题');"
`;

exports[`#transform > should transform code 1`] = `
"import { useI18n } from 'vue-i18n';
const t = useI18n();
const title = t('Key of 标题{0}', [msg]);"
`;

exports[`#transform > should transform code in function 1`] = `
"import fs from 'fs';
import { useI18n } from 'vue-i18n';
const t = useI18n();
let title = console.log(t('Key of 标题'));"
`;

exports[`#transform > should transform code in template 1`] = `
"import { useI18n } from 'vue-i18n';
const t = useI18n();
const title = t('Key of 标题{msg}', {
  msg: msg
});"
`;

exports[`#transform > should transform code when import 1`] = `
"import { useI18n } from 'vue-i18n';
const t = useI18n();
let x = t('s');
let title = t('Key of 标题');"
`;

exports[`#transform > should transform jsx code 1`] = `
"import { useI18n } from 'vue-i18n';
const t = useI18n();
const title = <div>{t('Key of 不好不好')}{aa}</div>;"
`;
